### ISSUE TITLE
Add Llama.cpp integration

### ISSUE THREAD

**Issue by davidmigloz** [davidmigloz/langchain_dart#317]

**Comment by danemadsen** [comment]
I was helping develop llama_cpp_dart for a while but i was having too many issues with it so i ended up creating my own which works alot better.

https://github.com/Mobile-Artificial-Intelligence/maid_llm

If you want to integrate that into langchain you should start there

**Comment by davidmigloz** [comment]
Ah I didn't know you dropped `llama_cpp_dart`, I'll check `maid_llm` out. Thanks for bringing this up.

If we integrate llama.cpp into LangChain.dart, I guess you could simplify `maid` implementation a bit, as all the providers will use the same standard interface LangChain.dart provides.

After I'm done with the [v0.5.0 release](https://github.com/davidmigloz/langchain_dart/milestone/28?closed=1), I'll dig into it.

**Comment by danemadsen** [comment]
Yeah, I did actually have it using langchains `ChatMessage` at one point but I changed it due to a few little things I wanted from my own `ChatNode` types.

llama_cpp_dart was a decent implementation, its probably the fastest implementation I've used thusfar but I was running into a few different sampling issues I simply couldn't fix in that implementation so I just ended up creating my own from scratch.

I've got vulkan acceleration working in my own now so yeah its abit better

With regards,
Dane Madsen
________________________________

**Comment by davidmigloz** [comment]
Good to know. If you miss any functionality in the current abstractions let me know, and we can add it. 

v0.5.0 [breaks](https://github.com/davidmigloz/langchain_dart/issues/328) `langchain` into `langchain_core`, `langchain` and `langchain_community`, so you could depend just on `langchain_core` if you only need to use the base abstractions without depending on the entire `langchain` package.

**Comment by danemadsen** [comment]
> Good to know. If you miss any functionality in the current abstractions let me know, and we can add it. 
> 
> v0.5.0 [breaks](https://github.com/davidmigloz/langchain_dart/issues/328) `langchain` into `langchain_core`, `langchain` and `langchain_community`, so you could depend just on `langchain_core` if you only need to use the base abstractions without depending on the entire `langchain` package.

Nah its probably nothing that really needs to be added into langchain. Its just that I'm using a tree structure for my chat messages so I can have branching conversations. You can add that in if you want it but your chatmessages are fine as is for most cases.

**Comment by raymon-io** [comment]
Is there any update on this? If there is some documentation for `maid_llm` , I could try to help integrate it into LangChain.

**Comment by davidmigloz** [comment]
@raymon-io I haven't started working on it yet, so it'd be great if you pick it up 🙂

We may have to create a `langchain_community_flutter` package to add integrations that depend on the Flutter SDK (as the current `langchain_community` package is a pure Dart one).

Happy to discuss it if you are interested!

**Comment by davidmigloz** [comment]
https://github.com/Telosnex/fllama

**Comment by breitburg** [comment]
Hey everyone!

I've made a bridge between [Fllama](https://github.com/Telosnex/fllama/) and `langchain_dart`. It's called [`langchain_fllama`](https://github.com/breitburg/langchain_fllama). I was able to make all the basic functionality of langchain working.

Link: https://github.com/breitburg/langchain_fllama

**Comment by danemadsen** [comment]
fllama is GPL so you probably shouldn't use it as it is incompatible with the MIT licence.

Ive kept `maid_llm` updated but I've renamed it to `llama_sdk` now. It auto updated llama.cpp daily now so that's helpful.

https://github.com/Mobile-Artificial-Intelligence/llama_sdk
